src/object_.d

```diff
@@ -153,26 +153,32 @@ class Object
 /************************
  * Returns true if lhs and rhs are equal.
  */
-bool opEquals(const Object lhs, const Object rhs)
-{
-    // A hack for the moment.
-    return opEquals(cast()lhs, cast()rhs);
-}
-
-bool opEquals(Object lhs, Object rhs)
+bool opEquals(T,U)(T lhs, U rhs)
 {
     // If aliased to the same object or both null => equal
-    if (lhs is rhs) return true;
+    static if(__traits(compiles, lhs is rhs)) {
+        if(lhs is rhs) return true;
+    }
 
-    // If either is null => non-equal
-    if (lhs is null || rhs is null) return false;
+    // If lhs is null, prevent calling a method on it by returning here
+    static if(__traits(compiles, lhs is null || rhs is null)) {
+        if(lhs is null || rhs is null) return false;
+    } else static if(__traits(compiles, lhs is null)) {
+        if(lhs is null) return false;
+    } else static if(__traits(compiles, rhs is null)) {
+        if(rhs is null) return false;
+    }
 
-    // If same exact type => one call to method opEquals
-    if (typeid(lhs) is typeid(rhs) || typeid(lhs).opEquals(typeid(rhs)))
+    static if( is ( typeof(cast()lhs) == typeof(cast()rhs) ) ) {
         return lhs.opEquals(rhs);
```

Comment:
No, in the general case we have to make two calls to `opEquals` methods: `lhs.opEquals(rhs)` and `rhs.opEquals(lhs)`, so that both objects have to agree to be equal. When the dynamic types are the same, only `lhs.opEquals(rhs)` is to be called. It's documented this way: http://dlang.org/operatoroverloading.html#equals (see the sample implementation of opEquals).